### PR TITLE
extend DynamicBucketDataset and add methods for bucketing

### PR DESCRIPTION
**Pull request by JanekEbb**
add argument total_size_threshold to DynamicBucketDataset allowing to have a smaller batch_size for long sequences (this can be useful to not exceed gpu memory). Added methods for (dynamic and precomputed) bucketing.

### CODE REVIEW

File: lazy_dataset/core.py [hunk 1]
```diff
@@ -1595,22 +1684,49 @@ class DynamicBucketDataset(Dataset):
...
+                can only be in a bucket with examples with lengths from 80 to 125.
+            total_size_threshold: total size of a bucket (len(bucket)*max_length_in_bucket)
+                after which a bucket is emitted though len(bucket) < batch_size
+            expiration: maximum life time of a bucket. After this number of
+                steps it is either emitted (if drop_incomplete is False)
+                else discarded
+            drop_incomplete: if True drop incomplete buckets at the end of
+                iteration or when buckets expire, else emit them.
+            sort_by_key: if True sort the bucket by the lengths of the examples
+                before emitting
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Could you mention that this is reverse sorting and explain the use-case?

File: lazy_dataset/core.py [hunk 2]
```diff
@@ -1640,26 +1757,69 @@ def indexable(self):
...
+                        len(bucket) >= self.batch_size
+                        or (
+                            self.max_total_size is not None
+                            and (len(bucket) * max_value) > self.max_total_size
+                        )
+                    ):
+                        expired.add(j)
+                        if self.sort_by_key:
+                            bucket = sorted(bucket, key=self.key, reverse=True)
                         yield bucket
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Is here a break missing?

File: lazy_dataset/core.py [hunk 3]
```diff
@@ -1640,26 +1757,69 @@ def indexable(self):
     def __iter__(self):
         buckets = list()
         for i, sample in enumerate(self.input_dataset):
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Can you change `sample` to `example`?

File: lazy_dataset/core.py [hunk 4]
```diff
@@ -1595,22 +1684,49 @@ class DynamicBucketDataset(Dataset):
...
+                max_total_size is set)
+            key: callable or dict key returning a scalar length given an
+                example dict
+            max_padding_rate: the maximum padding that has to be added to a
+                signal in a bucket. E.g. if set to 0.2, a example of length 100
+                can only be in a bucket with examples with lengths from 80 to 125.
+            total_size_threshold: total size of a bucket (len(bucket)*max_length_in_bucket)
+                after which a bucket is emitted though len(bucket) < batch_size
+            expiration: maximum life time of a bucket. After this number of
+                steps it is either emitted (if drop_incomplete is False)
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Can you explain what `steps` mean?

File: lazy_dataset/core.py [hunk 5]
```diff
@@ -1640,26 +1757,69 @@ def indexable(self):
...
             found_bucket = False
-            for i, (bucket, min_value, max_value) in enumerate(buckets):
-                if min_value <= value <= max_value:
+            expired = set()
+            for j, (
+                bucket, creation_idx, lower_bound, upper_bound, max_value
+            ) in enumerate(buckets):
+                if (
+                    self.expiration is not None
+                    and (i - creation_idx) / self.batch_size > self.expiration
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Can you write a comment, why you do not count the yielded batches?
`batch_size` is only correct, when no batch was yielded incomplete or dropped.

**Comment by boeddeker** [comment]
The function `get_bucket_indices` has some problems.
It does not check if the input is indexable (an assert would solve this).
But more important: It can silently produce wrong batches because an intermediate shuffle could change the order.

```python
>>> import lazy_dataset
>>> examples = {'a': {'x': 1}, 'b': {'x': 3},  'c': {'x': 12}, 'd': {'x': 2}}
>>> it = lazy_dataset.core.DictDataset(examples)
>>> bucket_indices = it.get_bucket_indices('x', 2)
>>> list(it.shuffle().batch_bucket(bucket_indices, 2))  # silently produces the wrong output
[[{'x': 12}, {'x': 2}], [{'x': 3}, {'x': 1}]]
>>> it.shuffle(reshuffle=True).get_bucket_indices('x', 2)  # should not work
array([1, 0, 0, 1])
```

Maybe `get_bucket_indices` could return a dict/orderddict with the mapping key to bucket index.
This would limit `get_bucket_indices` to datasets that have keys, but the order at that position does not matter.

It would be nice, when the following examples would work:
```python
>>> it.shuffle(reshuffle=True).batch_bucket(...)
>>> it.shuffle(reshuffle=True).batch_bucket(...).prefetch(...)
```

**Comment by boeddeker** [comment]
Can you add `numpy>=1.15.0` as a dependency? See https://docs.scipy.org/doc/numpy/reference/generated/numpy.quantile.html

File: lazy_dataset/core.py (hunk 2 above)
Comment (JanekEbb) [reply lazy_dataset/core.py]: I keep iterating to yield buckets which have exceeded total_size_threshold

Comment (boeddeker) [reply lazy_dataset/core.py]: Ah, my fault, I overread the `elif not found_bucket`

File: lazy_dataset/core.py [hunk 6]
```diff
@@ -1640,26 +1757,69 @@ def indexable(self):
...
+                ):
+                    expired.add(j)
+                    if not self.drop_incomplete:
+                        if self.sort_by_key:
+                            bucket = sorted(bucket, key=self.key, reverse=True)
+                        yield bucket
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Is it intended, that the "sample" cannot be appended to expired buckets? (`lower_bound <= value <= upper_bound` can be true for an `expired` bucket)

File: lazy_dataset/core.py [hunk 7]
```diff
@@ -1851,28 +1911,77 @@ def indexable(self):
...
+            ) in enumerate(buckets):
+                if lower_bound <= value <= upper_bound:
+                    bucket.append(example)
+                    max_value = max(max_value, value)
+                    if (
+                            len(bucket) >= self.batch_size
+                            or (
+                            self.total_size_threshold is not None
+                            and (len(
+                        bucket) * max_value) > self.total_size_threshold
```
Comment (boeddeker) [inline lazy_dataset/core.py]: When `total_size_threshold` is used, the batch if yielded, when `len(bucket)*max_length_in_bucket > total_size_threshold`. Can you document, that the batch size is bigger than the threshold?
In extreme cases it is approximately `2 * max_length_in_bucket`.

**Pull request merged**